// ng-gallery/esm2022/lib/utils/img-manager.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject, filter, map, switchMap, EMPTY } from 'rxjs';
import * as i0 from "@angular/core";
export class ImgManager {
    constructor() {
        this.trigger$ = new BehaviorSubject(null);
        this.images = new Map();
    }
    getActiveItem(state$) {
        return this.trigger$.pipe(switchMap(() => state$.pipe(switchMap((state) => {
            const img = this.images.get(state.currIndex);
            if (img) {
                return img.state.pipe(filter((state) => state !== 'loading'), map(() => img.target));
            }
            return EMPTY;
        }))));
    }
    addItem(index, payload) {
        this.images.set(index, payload);
        this.trigger$.next();
    }
    deleteItem(index) {
        if (this.images.has(index)) {
            this.images.delete(index);
            this.trigger$.next();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ImgManager, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ImgManager }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ImgManager, decorators: [{
            type: Injectable
        }] });
