// ng-gallery/esm2022/lib/utils/img-recognizer.mjs
import { Directive, Input, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./img-manager";
import * as i2 from "../components/gallery-item.component";
export class ImgRecognizer {
    onLoad() {
        this.item.state$.next('success');
    }
    onError() {
        this.item.state$.next('failed');
    }
    constructor(el, manager, item) {
        this.el = el;
        this.manager = manager;
        this.item = item;
        if (item) {
            // Mark the gallery-item component as an image item
            item.isItemContainImage = true;
        }
        else {
            throw new Error('[NgGallery]: galleryImage directive should be only used inside gallery item templates!');
        }
    }
    ngOnInit() {
        this.manager.addItem(this.index, {
            state: this.item.state$.asObservable(),
            target: this.el.nativeElement
        });
    }
    ngOnDestroy() {
        this.manager.deleteItem(this.index);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ImgRecognizer, deps: [{ token: i0.ElementRef }, { token: i1.ImgManager }, { token: i2.GalleryItemComponent }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.1", type: ImgRecognizer, isStandalone: true, selector: "img[galleryImage]", inputs: { index: ["galleryImage", "index"] }, host: { listeners: { "load": "onLoad($event)", "error": "onError($event)" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ImgRecognizer, decorators: [{
            type: Directive,
            args: [{
                    selector: 'img[galleryImage]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i1.ImgManager }, { type: i2.GalleryItemComponent }], propDecorators: { index: [{
                type: Input,
                args: ['galleryImage']
            }], onLoad: [{
                type: HostListener,
                args: ['load', ['$event']]
            }], onError: [{
                type: HostListener,
                args: ['error', ['$event']]
            }] } });
