// ng-gallery/esm2022/lib/autoplay/autoplay.directive.mjs
import { Directive, Input } from '@angular/core';
import { delay, of, switchMap, tap, EMPTY } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "../services/gallery.service";
import * as i2 from "../utils/img-manager";
export class AutoplayDirective {
    constructor(_gallery, _imgManager) {
        this._gallery = _gallery;
        this._imgManager = _imgManager;
    }
    ngAfterViewInit() {
        this._galleryRef = this._gallery.ref(this.galleryId);
        this._subscribe();
        if (this.config.autoplay) {
            this._galleryRef.play();
        }
    }
    ngOnChanges(changes) {
        if (this._galleryRef && changes.config?.currentValue.autoplay !== changes.config?.previousValue.autoplay) {
            this.config.autoplay ? this._galleryRef.play() : this._galleryRef.stop();
        }
    }
    ngOnDestroy() {
        this._unsubscribe();
    }
    _subscribe() {
        this._unsubscribe();
        this._currentSubscription = this._galleryRef.playingChanged.pipe(switchMap((state) => {
            if (state.isPlaying) {
                return this._imgManager.getActiveItem(this._galleryRef.state).pipe(switchMap(() => of({}).pipe(delay(this.config.autoplayInterval), tap(() => {
                    if (this._galleryRef.stateSnapshot.hasNext) {
                        this._galleryRef.next(this.config.scrollBehavior);
                    }
                    else {
                        this._galleryRef.set(0, this.config.scrollBehavior);
                    }
                }))));
            }
            return EMPTY;
        })).subscribe();
    }
    _unsubscribe() {
        this._currentSubscription?.unsubscribe();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: AutoplayDirective, deps: [{ token: i1.Gallery }, { token: i2.ImgManager }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.1", type: AutoplayDirective, isStandalone: true, selector: "gallery-core[autoplay]", inputs: { config: "config", galleryId: "galleryId" }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: AutoplayDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'gallery-core[autoplay]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i1.Gallery }, { type: i2.ImgManager }], propDecorators: { config: [{
                type: Input
            }], galleryId: [{
                type: Input
            }] } });
