// ng-gallery/esm2022/lib/observers/item-intersection-observer.directive.mjs
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { combineLatest, filter, switchMap } from 'rxjs';
import { ActiveItemObserver } from './active-item-observer';
import { resizeObservable } from '../utils/resize-observer';
import * as i0 from "@angular/core";
import * as i1 from "../components/gallery-item.component";
export class ItemIntersectionObserver {
    get _viewport() {
        return this._item.nativeElement.parentElement.parentElement;
    }
    constructor(_zone, _item) {
        this._zone = _zone;
        this._item = _item;
        this._sensor = new ActiveItemObserver();
        this.activeIndexChange = new EventEmitter();
    }
    ngOnChanges() {
        (this.config.itemAutosize && !this.disabled) ? this._subscribe() : this._unsubscribe();
    }
    ngOnDestroy() {
        this._unsubscribe();
    }
    _subscribe() {
        this._unsubscribe();
        if (!!this.adapter) {
            this._zone.runOutsideAngular(() => {
                this._currentSubscription = combineLatest([
                    resizeObservable(this._viewport),
                    resizeObservable(this._item.nativeElement)
                ]).pipe(switchMap(() => this._item.state$), filter((state) => state !== 'loading'), switchMap(() => {
                    const rootMargin = this.adapter.getElementRootMargin(this._viewport, this._item.nativeElement);
                    if (this.config.debug) {
                        this._item.nativeElement.style.setProperty('--item-intersection-margin', `"VIEWPORT(${this._viewport.clientWidth}x${this._viewport.clientHeight}) ITEM(${this._item.nativeElement.clientWidth}x${this._item.nativeElement.clientHeight}) INTERSECTION(${rootMargin})"`);
                    }
                    return this._sensor.observe(this._viewport, [this._item.nativeElement], rootMargin);
                })).subscribe((index) => {
                    this._zone.run(() => this.activeIndexChange.emit(index));
                });
            });
        }
    }
    _unsubscribe() {
        this._currentSubscription?.unsubscribe();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ItemIntersectionObserver, deps: [{ token: i0.NgZone }, { token: i1.GalleryItemComponent }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.1", type: ItemIntersectionObserver, isStandalone: true, selector: "[itemIntersectionObserver]", inputs: { adapter: "adapter", config: "config", disabled: ["itemIntersectionObserverDisabled", "disabled"] }, outputs: { activeIndexChange: "activeIndexChange" }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ItemIntersectionObserver, decorators: [{
            type: Directive,
            args: [{
                    selector: '[itemIntersectionObserver]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.NgZone }, { type: i1.GalleryItemComponent }], propDecorators: { adapter: [{
                type: Input
            }], config: [{
                type: Input
            }], disabled: [{
                type: Input,
                args: ['itemIntersectionObserverDisabled']
            }], activeIndexChange: [{
                type: Output
            }] } });
