// ng-gallery/esm2022/lib/observers/slider-intersection-observer.directive.mjs
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { ActiveItemObserver } from './active-item-observer';
import * as i0 from "@angular/core";
export class SliderIntersectionObserver {
    get _viewport() {
        return this._el.nativeElement;
    }
    constructor(_zone, _el) {
        this._zone = _zone;
        this._el = _el;
        this._sensor = new ActiveItemObserver();
        this.activeIndexChange = new EventEmitter();
    }
    ngOnChanges() {
        (this.config.itemAutosize || this.disabled) ? this._unsubscribe() : this._subscribe();
    }
    ngOnDestroy() {
        this._unsubscribe();
    }
    _subscribe() {
        this._unsubscribe();
        if (!!this.adapter && !!this.items?.length) {
            const rootMargin = this.adapter.getRootMargin();
            if (this.config.debug) {
                this._viewport.style.setProperty('--intersection-margin', `"INTERSECTION(${rootMargin})"`);
            }
            this._zone.runOutsideAngular(() => {
                this._currentSubscription = this._sensor.observe(this._viewport, this.items.map((item) => item.nativeElement), rootMargin).subscribe((index) => {
                    this._zone.run(() => this.activeIndexChange.emit(index));
                });
            });
        }
    }
    _unsubscribe() {
        this._currentSubscription?.unsubscribe();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: SliderIntersectionObserver, deps: [{ token: i0.NgZone }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.1", type: SliderIntersectionObserver, isStandalone: true, selector: "[sliderIntersectionObserver]", inputs: { adapter: "adapter", items: "items", config: "config", disabled: ["sliderIntersectionObserverDisabled", "disabled"] }, outputs: { activeIndexChange: "activeIndexChange" }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: SliderIntersectionObserver, decorators: [{
            type: Directive,
            args: [{
                    selector: '[sliderIntersectionObserver]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.NgZone }, { type: i0.ElementRef }], propDecorators: { adapter: [{
                type: Input
            }], items: [{
                type: Input
            }], config: [{
                type: Input
            }], disabled: [{
                type: Input,
                args: ['sliderIntersectionObserverDisabled']
            }], activeIndexChange: [{
                type: Output
            }] } });
