// ng-gallery/esm2022/lib/observers/thumb-resize-observer.directive.mjs
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { tap, animationFrameScheduler, debounceTime } from 'rxjs';
import { resizeObservable } from '../utils/resize-observer';
import * as i0 from "@angular/core";
export class ThumbResizeObserver {
    get _viewport() {
        return this._el.nativeElement;
    }
    constructor(_el, _zone) {
        this._el = _el;
        this._zone = _zone;
        this.resized = new EventEmitter();
    }
    ngOnInit() {
        this._zone.runOutsideAngular(() => {
            this._resizeSubscription = resizeObservable(this._viewport).pipe(debounceTime(this.config.resizeDebounceTime, animationFrameScheduler), tap(() => {
                this.updateSliderSize();
                this.resized.emit();
            })).subscribe();
        });
    }
    ngOnChanges(changes) {
        if (!changes.config.firstChange) {
            this.updateSliderSize();
        }
    }
    ngOnDestroy() {
        this._resizeSubscription?.unsubscribe();
    }
    updateSliderSize() {
        this._viewport.style.setProperty('--thumb-centralize-start-size', this.adapter.getCentralizerStartSize() + 'px');
        this._viewport.style.setProperty('--thumb-centralize-end-size', this.adapter.getCentralizerEndSize() + 'px');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ThumbResizeObserver, deps: [{ token: i0.ElementRef }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.1", type: ThumbResizeObserver, isStandalone: true, selector: "[thumbResizeObserver]", inputs: { config: "config", adapter: "adapter" }, outputs: { resized: "thumbResizeObserver" }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: ThumbResizeObserver, decorators: [{
            type: Directive,
            args: [{
                    selector: '[thumbResizeObserver]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i0.NgZone }], propDecorators: { config: [{
                type: Input
            }], adapter: [{
                type: Input
            }], resized: [{
                type: Output,
                args: ['thumbResizeObserver']
            }] } });
