// paychangu-js/dist/services/level.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LevelAPI = void 0;
const errors_1 = require("../utils/errors");
const validation_1 = require("../utils/validation");
const helpers_1 = require("../utils/helpers");
class LevelAPI {
    constructor() {
        this.secretKey = null;
    }
    static getInstance() {
        if (!LevelAPI.instance) {
            LevelAPI.instance = new LevelAPI();
        }
        return LevelAPI.instance;
    }
    setSecretKey(secretKey) {
        this.secretKey = secretKey;
    }
    initiateTransaction(config) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.secretKey) {
                throw new errors_1.PaychanguError('Secret key not set. Call setSecretKey() first.');
            }
            (0, validation_1.validateLevelConfig)(config);
            return (0, helpers_1.retry)(() => __awaiter(this, void 0, void 0, function* () {
                try {
                    const response = yield fetch('https://api.paychangu.com/payment', {
                        method: 'POST',
                        headers: {
                            'Accept': 'application/json',
                            'Authorization': `Bearer ${this.secretKey}`,
                            'Content-Type': 'application/json',
                        },
                        body: JSON.stringify(config),
                    });
                    if (!response.ok) {
                        throw new errors_1.PaychanguError(`HTTP error! status: ${response.status}`);
                    }
                    const data = yield response.json();
                    return data;
                }
                catch (error) {
                    if (error instanceof errors_1.PaychanguError) {
                        throw error;
                    }
                    throw new errors_1.PaychanguError('Failed to initiate transaction', { cause: error });
                }
            }), 3, 1000);
        });
    }
}
exports.LevelAPI = LevelAPI;
