// paychangu-js/dist/services/verification.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationAPI = void 0;
const errors_1 = require("../utils/errors");
const helpers_1 = require("../utils/helpers");
class VerificationAPI {
    constructor() {
        this.secretKey = null;
    }
    static getInstance() {
        if (!VerificationAPI.instance) {
            VerificationAPI.instance = new VerificationAPI();
        }
        return VerificationAPI.instance;
    }
    setSecretKey(secretKey) {
        this.secretKey = secretKey;
    }
    verifyTransaction(txRef) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.secretKey) {
                throw new errors_1.PaychanguError('Secret key not set. Call setSecretKey() first.');
            }
            return (0, helpers_1.retry)(() => __awaiter(this, void 0, void 0, function* () {
                try {
                    const response = yield fetch(`https://api.paychangu.com/verify-payment/${txRef}`, {
                        method: 'GET',
                        headers: {
                            'Accept': 'application/json',
                            'Authorization': `Bearer ${this.secretKey}`,
                        },
                    });
                    if (!response.ok) {
                        throw new errors_1.PaychanguError(`HTTP error! status: ${response.status}`);
                    }
                    const data = yield response.json();
                    return data;
                }
                catch (error) {
                    if (error instanceof errors_1.PaychanguError) {
                        throw error;
                    }
                    throw new errors_1.PaychanguError('Failed to verify transaction', { cause: error });
                }
            }), 3, 1000);
        });
    }
}
exports.VerificationAPI = VerificationAPI;
