// ng-gallery/esm2022/lib/components/templates/items.model.mjs
import { GalleryItemTypes } from '../../models/constants';
export class ImageItem {
    constructor(data) {
        this.data = data;
        this.type = GalleryItemTypes.Image;
    }
}
export class VideoItem {
    constructor(data) {
        this.data = data;
        this.type = GalleryItemTypes.Video;
    }
}
export class IframeItem {
    constructor(data) {
        this.data = data;
        this.type = GalleryItemTypes.Iframe;
    }
}
export class YoutubeItem {
    constructor(data) {
        this.data = {
            ...data,
            ...{
                src: `https://youtube.com/embed/${data.src}`,
                thumb: data.thumb ? data.thumb : `//img.youtube.com/vi/${data.src}/default.jpg`
            }
        };
        this.type = GalleryItemTypes.Youtube;
    }
}
export class VimeoItem {
    constructor(data) {
        this.data = {
            ...data,
            ...{
                src: `https://player.vimeo.com/video/${data.src}`,
                thumb: data.thumb ? data.thumb : this.getVimeoThumb(data.src)
            }
        };
        this.type = GalleryItemTypes.Vimeo;
    }
    getVimeoThumb(videoId) {
        //Vimeo has no API for getting a thumbnail, but this project can do it: https://github.com/ThatGuySam/vumbnail
        return `//vumbnail.com/${videoId}.jpg`;
    }
}
