// ng-gallery/esm2022/lib/components/adapters/main-adapters.mjs
/**
 * A clone of HammerJs constants
 */
export const DIRECTION_LEFT = 2;
export const DIRECTION_RIGHT = 4;
export const DIRECTION_UP = 8;
export const DIRECTION_DOWN = 16;
export class HorizontalAdapter {
    get scrollValue() {
        return this.slider.scrollLeft;
    }
    get clientSize() {
        return this.slider.clientWidth;
    }
    get isContentLessThanContainer() {
        return this.clientSize >= this.slider.firstElementChild.clientWidth;
    }
    constructor(slider, config) {
        this.slider = slider;
        this.config = config;
        this.hammerDirection = DIRECTION_LEFT | DIRECTION_RIGHT;
        this.scrollSnapType = 'x mandatory';
    }
    getScrollToValue(el, behavior) {
        const position = el.offsetLeft - ((this.clientSize - el.clientWidth) / 2);
        return {
            behavior,
            start: position
        };
    }
    getRootMargin() {
        return `1000px 1px 1000px 1px`;
    }
    getElementRootMargin(viewport, el) {
        const rootMargin = -1 * ((viewport.clientWidth - el.clientWidth) / 2) + 1;
        return `0px ${rootMargin}px 0px ${rootMargin}px`;
    }
    getCentralizerStartSize() {
        if (this.isContentLessThanContainer) {
            const size = this.clientSize - this.slider.firstElementChild.clientWidth;
            return size / 2;
        }
        return (this.clientSize / 2) - (this.slider.firstElementChild.firstElementChild?.clientWidth / 2);
    }
    getCentralizerEndSize() {
        if (this.isContentLessThanContainer) {
            const size = this.clientSize - this.slider.firstElementChild.clientWidth;
            return size / 2;
        }
        return (this.clientSize / 2) - (this.slider.firstElementChild.lastElementChild?.clientWidth / 2);
    }
    getHammerVelocity(e) {
        return e.velocityX;
    }
    getHammerValue(value, e, behavior) {
        return {
            behavior,
            left: value - e.deltaX
        };
    }
}
export class VerticalAdapter {
    get scrollValue() {
        return this.slider.scrollTop;
    }
    get clientSize() {
        return this.slider.clientHeight;
    }
    get isContentLessThanContainer() {
        return this.clientSize >= this.slider.firstElementChild.clientHeight;
    }
    constructor(slider, config) {
        this.slider = slider;
        this.config = config;
        this.hammerDirection = DIRECTION_UP | DIRECTION_DOWN;
        this.scrollSnapType = 'y mandatory';
    }
    getScrollToValue(el, behavior) {
        const position = el.offsetTop - ((this.clientSize - el.clientHeight) / 2);
        return {
            behavior,
            top: position
        };
    }
    getRootMargin() {
        return `1px 1000px 1px 1000px`;
    }
    getElementRootMargin(viewport, el) {
        const rootMargin = -1 * ((viewport.clientHeight - el.clientHeight) / 2) + 1;
        return `${rootMargin}px 0px ${rootMargin}px 0px`;
    }
    getCentralizerStartSize() {
        if (this.isContentLessThanContainer) {
            const size = this.clientSize - this.slider.firstElementChild.clientHeight;
            return size / 2;
        }
        return (this.clientSize / 2) - (this.slider.firstElementChild.firstElementChild?.clientHeight / 2);
    }
    getCentralizerEndSize() {
        if (this.isContentLessThanContainer) {
            const size = this.clientSize - this.slider.firstElementChild.clientHeight;
            return size / 2;
        }
        return (this.clientSize / 2) - (this.slider.firstElementChild.lastElementChild?.clientHeight / 2);
    }
    getHammerVelocity(e) {
        return e.velocityY;
    }
    getHammerValue(value, e, behavior) {
        return {
            behavior,
            top: value - e.deltaY
        };
    }
}
